### ISSUE TITLE
FR: Tailscale based Anycast IP address

### ISSUE THREAD

**Issue by charlieporth1** [tailscale/tailscale#3860]
### What are you trying to do?

Tailscale based Anycast IP address. The use case is for my private DNS servers. 

### How should we solve this?

I don't know if Wireguard supports this directly but I would love any input on this idea

### What is the impact of not solving this?

Anycast provides great uptime and reliability among servers 

### Anything else?
https://www.cloudflare.com/learning/cdn/glossary/anycast-network/

**Comment by DentonGentry** [comment]
Do you have a specific case you'd like to solve using anycast IP addresses?

Versus DNS-based solutions for example, supplying a different IP address in the Answer based on the location of the client making the request?

**Comment by charlieporth1** [comment]
Sorry that didn't clarify this. Anycast IP being for DNS servers. Not requests. The specific use case being I own these servers but I do not own the IP addresses or how they connect to the internet. I also own them in different locations and through different providers. Meaning anycast IP is next to impossible through a consumer standard. These DNS servers are for personal use none the less, but this would also provide better availability and reliability. Due to the nature of anycast. It's supposed to be with client requests in order to do with server IPs

**Comment by charlieporth1** [comment]
By the way congratulations on your funding!!!!

**Comment by DentonGentry** [comment]
Regarding DNS servers, let's say you have four DNS servers spaced around the planet:
100.101.101.233
100.101.101.234
100.101.101.235
100.101.101.236

If you add all four of these as Global Nameservers in https://login.tailscale.com/admin/dns, they will all be pushed in the DNS config to the MagicDNS service running on each of your devices.

MagicDNS sends queries to all configured nameservers in parallel (*), and uses whichever one responds first. So the query should reach the closest DNS server first and get a response back in roughly the optimal time, without needing Anycast. The rate of DNS queries generated by even a large tailnet is still quite manageable in this way, it tends not to overload any modern DNS server.

(*) actually it knows that 8.8.8.8 and 8.8.4.4 and the IPv6 addresses are all Google DNS, and only sends to one of them not all four in parallel. It also knows about CloudFlare's addresses, and a few other popular providers.

**Comment by charlieporth1** [comment]
Interesting work around. Is there any advantage to using Anycast over that?

**Comment by DentonGentry** [comment]
I don't really understand the point about configuring different DNS servers per client, but it isn't necessary for the parallel queries described in https://github.com/tailscale/tailscale/issues/3860#issuecomment-1027545909. All clients can be given all configured DNS servers addresses, they will send to all of them in parallel, and use the first which responds.

I say this because it is unlikely that Tailscale will implement endpoint management so complex as to hand out different DNS server configurations for subsets of the clients.

**Comment by bradfitz** [comment]
It's likely we won't do this until we flesh out our "services" & load balancing more. We have the start of a service map and the start of HA subnet routers at least, but we don't currently let you run active/active HA services. When we do, it'll probably be implemented by Tailscale IPs that are anycast-like: magicsock will route them to whichever node offering that service is nearest.

**Comment by DentonGentry** [comment]
I'm going to close this as I think adding all of the DNS servers will do what is requested without needing an Anycast function. Client devices will race all of the DNS servers and use the first one which responds. Please reopen if not.

This results in extra DNS lookups sent to each server. One would have to have a very large number of clients before that load would be problematic.
